lib/ThreeElementVectorLoweringPass.cpp

```diff
@@ -0,0 +1,1419 @@
+  ValueMap.insert({&U, &V});
+
+  if (U.getType() == V.getType()) {
+    LLVM_DEBUG(dbgs() << "\tAnd replace its usages.\n");
+    U.replaceAllUsesWith(&V);
+  }
+
+  if (U.hasName()) {
+    V.takeName(&U);
+  }
+
+  auto *I = dyn_cast<Instruction>(&U);
+  auto *J = dyn_cast<Instruction>(&V);
+  if (I && J) {
+    J->copyMetadata(*I);
+  }
+}
+
+Type *ThreeElementVectorLoweringPass::getEquivalentType(Type *Ty) {
+  auto it = TypeMap.find(Ty);
+  if (it != TypeMap.end()) {
+    return it->second;
+  }
+
+  // Recursive implementation, taking advantage of the cache.
+  auto *EquivalentTy = getEquivalentTypeImpl(Ty);
+  TypeMap.insert({Ty, EquivalentTy});
+
+  if (EquivalentTy) {
+    LLVM_DEBUG(dbgs() << "Generating equivalent type for " << *Ty << ": "
+                      << *EquivalentTy << '\n');
+  }
+
+  return EquivalentTy;
+}
+
+Type *ThreeElementVectorLoweringPass::getEquivalentTypeImpl(Type *Ty) {
+  if (Ty->isIntegerTy() || Ty->isFloatingPointTy() || Ty->isVoidTy() ||
+      Ty->isLabelTy() || Ty->isMetadataTy()) {
+    // No lowering required.
+    return nullptr;
+  }
+
+  if (auto *VectorTy = dyn_cast<VectorType>(Ty)) {
+    unsigned Arity = VectorTy->getElementCount().getKnownMinValue();
+    bool RequireLowering = (Arity == 3);
+
+    if (RequireLowering) {
+      assert(!VectorTy->getElementCount().isScalable() &&
+             "Unsupported scalable vector");
+
+      // This assumes that the element type of the vector is a primitive scalar.
+      // That is, no vectors of pointers for example.
+      Type *ScalarTy = VectorTy->getElementType();
+      assert((ScalarTy->isFloatingPointTy() || ScalarTy->isIntegerTy()) &&
+             "Unsupported scalar type");
+
+      return VectorType::get(ScalarTy, 4, false);
+    }
+
+    return nullptr;
+  }
+
+  if (auto *PointerTy = dyn_cast<PointerType>(Ty)) {
+    if (auto *ElementTy = getEquivalentType(PointerTy->getElementType())) {
+      return ElementTy->getPointerTo(PointerTy->getAddressSpace());
+    }
+
+    return nullptr;
+  }
+
+  if (auto *ArrayTy = dyn_cast<ArrayType>(Ty)) {
+    if (auto *ElementTy = getEquivalentType(ArrayTy->getElementType())) {
+      return ArrayType::get(ElementTy, ArrayTy->getNumElements());
+    }
+
+    return nullptr;
+  }
+
+  if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+    unsigned Arity = StructTy->getStructNumElements();
+    if (Arity == 0)
+      return nullptr;
+    LLVMContext &Ctx = StructTy->getContainedType(0)->getContext();
+    SmallVector<Type *, 16> Types;
+    bool RequiredLowering = false;
+    for (unsigned i = 0; i < Arity; ++i) {
+      Type *CTy = StructTy->getContainedType(i);
+      auto *EquivalentTy = getEquivalentType(CTy);
+      if (EquivalentTy != nullptr) {
+        Types.push_back(EquivalentTy);
+        RequiredLowering = true;
+      } else {
+        Types.push_back(CTy);
+      }
+    }
+
+    if (RequiredLowering) {
+      return StructType::get(Ctx, Types, StructTy->isPacked());
```

Comment:
If the struct was packed, lowering from vec3 to vec4 might change the layout. I think the following can be packed into 16 bytes:

```
struct S {
  float3 x;
  int y;
};
```

If you have:
```
struct S {
  float4 x;
  int y;
};
```
The offset for y will be at 16 bytes making the total size 20 bytes.

Reply:
I agree.
It is not an easy problem.
I think a proper solution would be to say that float3 inside a packed structure is just not supported by clspv at the moment, and have a llvm_unreachable for that case. What do you think?
I am pushing another "solution" which do not lower packed structure. But I am not very confident about this implementation as it changes quite a lot in the way the pass works, and I did not test it enough.